iml-api/src/graphql.rs

```diff
@@ -478,6 +559,176 @@ impl QueryRoot {
 
         Ok(xs)
     }
+
+    #[graphql(arguments(
+        limit(description = "optional paging limit, defaults to all rows"),
+        offset(description = "Offset into items, defaults to 0"),
+        dir(description = "Sort direction, defaults to asc"),
+        message(description = "Pattern to search for in message"),
+        fqdn(description = "Pattern to search for in FQDN"),
+        tag(description = "Pattern to search for in tag"),
+        start_datetime(description = "Start of the time period of logs"),
+        end_datetime(description = "End of the time period of logs"),
+        message_class(description = "Array of log message classes"),
+        severity(description = "Upper bound of log severity"),
+    ))]
+    async fn logs(
+        context: &Context,
+        limit: Option<i32>,
+        offset: Option<i32>,
+        dir: Option<SortDir>,
+        message: Option<String>,
+        fqdn: Option<String>,
+        tag: Option<String>,
+        start_datetime: Option<String>,
+        end_datetime: Option<String>,
+        message_class: Vec<MessageClass>,
+        severity: Option<LogSeverity>,
+    ) -> juniper::FieldResult<Vec<LogMessage>> {
+        let dir = dir.unwrap_or_default();
+
+        let start_datetime: Option<chrono::DateTime<Utc>> = if let Some(s) = start_datetime {
+            Some(s.parse()?)
+        } else {
+            None
+        };
+        let end_datetime: Option<chrono::DateTime<Utc>> = if let Some(e) = end_datetime {
+            Some(e.parse()?)
+        } else {
+            None
+        };
+        let message_class: Vec<_> = if !message_class.is_empty() {
+            message_class
+        } else {
+            vec![MessageClass::Normal]
+        }
+        .into_iter()
+        .map(|i| i as i16)
+        .collect();
+        let severity = if let Some(s) = severity {
+            s as i16
+        } else {
+            LogSeverity::Debug as i16
+        };
+
+        let results = match (start_datetime, end_datetime) {
+            (Some(s), Some(e)) => {
+                sqlx::query_as!(
+                    LogMessageRecord,
+                    r#"
+                    SELECT * FROM chroma_core_logmessage t
+                    WHERE t.message LIKE $4
+                      AND t.fqdn LIKE $5
+                      AND t.tag LIKE $6
+                      AND t.datetime >= $7
+                      AND t.datetime < $8
```

Comment:
If we change it to 
```
AND ($7::TIMESTAMPTZ IS NULL OR t.datetime >= $7)
AND ($8::TIMESTAMPTZ IS NULL OR t.datetime < $8)
```
and pass options, this would allow you to remove the duplicated SQL snippets, no?
I don't strictly against the duplication, just trying to see, why it is used here.

Reply:
Nice, I'll try this